### ISSUE TITLE
debug info: flag functions as static

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#1860]
> rnk added a comment.
> 
> In D54667#1341110 <https://reviews.llvm.org/D54667#1341110>, @andrewrk wrote:
>> In Zig, a struct acts as a namespace full of static functions, so what you said works perfectly.
>>
>> I'm also open for suggestions if there was a better way to represent this in LLVM's debug info.

> In that case, you might want to add the DINode::FlagStaticMember flag to all your methods. If a 
> DISubprogram lists a DICompositeType as its scope or is used as an element of the composite type, 
> it gets treated as a C++-style method where the first argument is assumed to be 'this'. That's a very 
> C++-centric way to do things, so it might be nicer if we explicitly marked C++ instance methods, but 
> that's the way things are today and changing it would require thinking about backwards compatibility. 
> 
> The benefit of explicitly marking everything as a "static" method for zig is that if the first argument 
>happens to be a pointer type it won't end up looking like an instance method.

**Comment by vegecode** [comment]
```
!656 = distinct !DISubprogram(name: "assert", scope: !167, file: !167, line: 113, type: !657, isLocal: true, isDefini
tion: true, scopeLine: 113, flags: DIFlagStaticMember, isOptimized: false, unit: !1, retainedNodes: !660)
```
Is this what it should look like? How can I test this besides manually looking at the IR. This was just a one line change in ZigLLVMCreateFunction.

**Comment by kristate** [comment]
@vegecode 

Number of different ways to check.

First compile with COFF information (windows):
`zig test file.zig --target-os windows --target-arch x86_64`

then you can use `llvm-pdbutil` and `llvm-readobj`:
`llvm-pdbutil dump -types -dependents -type-data zig-cache/test.pdb`
`llvm-readobj -codeview zig-cache/test.obj`

Will check back later to dig deeper.

**Comment by andrewrk** [comment]
@vegecode I think that's what it should look like for functions whose scope is a struct but not global functions (although it may be harmless, depending on how LLVM handles it).

Ideally we would have a set of tests checking that the desired LLVM IR is emittted, and specifically for testing debug info, but we don't have that right now and I don't expect it as part of this issue.  So for now, hand verification of debug info / LLVM IR is all we have.

**Comment by vegecode** [comment]
My one line pull request. Is that ok for pull requests to be so small?
#1861 

I used the method from @kristate above to understand how the debug info was different. Thank you!

I tried to implement what @andrewrk said as far as global functions not having the flag set, but at the only call site for the function, I tried this to filter for functions within a struct: 

```
bool is_struct_member_function = (fn_scope->base.parent->source_node->type ==
                    NodeTypeContainerDecl);
```

Every function ended up with the flag set.

# Test Code

```rust
const std = @import("std");
const assert = std.debug.assert;

fn globalFunction() i32 {
    return 234;
}

const AddingStruct = struct {
    latest_result: i32,
    const Self = @This();

    fn add(self: *Self, a:i32, b:i32) i32 {
        const retVal = a + b;
        self.latest_result = retVal;
        return retVal;
    }
};

test "call struct static method" {
    var adding_struct = AddingStruct {
        .latest_result = undefined,
        };
    const result = adding_struct.add(1,2);
    assert(adding_struct.latest_result == 3);
    assert(result == 3);
    assert(globalFunction() == 234);
}
```

# With DIFlagStaticMember. Partial llvm-readobj output

```
  MemberFunction (0x1040) {
    TypeLeafKind: LF_MFUNCTION (0x1009)
    ReturnType: int (0x74)
    ClassType: AddingStruct (0x1025)
    ThisType: 0x0
    CallingConvention: NearC (0x0)
    FunctionOptions [ (0x0)
    ]
    NumParameters: 3
    ArgListType: (AddingStruct*, int, int) (0x103F)
    ThisAdjustment: 0
  }
  MemberFuncId (0x1041) {
    TypeLeafKind: LF_MFUNC_ID (0x1602)
    ClassType: AddingStruct (0x1025)
    FunctionType: int AddingStruct::(AddingStruct*, int, int) (0x1040)
    Name: AddingStruct_add
  }
```

# Without DIFlagStaticMember

```
  MemberFunction (0x1040) {
    TypeLeafKind: LF_MFUNCTION (0x1009)
    ReturnType: int (0x74)
    ClassType: AddingStruct (0x1025)
    ThisType: 0x0
    CallingConvention: NearC (0x0)
    FunctionOptions [ (0x0)
    ]
    NumParameters: 3
    ArgListType: (int, int) (0x103F)
    ThisAdjustment: 0
  }
  MemberFuncId (0x1041) {
    TypeLeafKind: LF_MFUNC_ID (0x1602)
    ClassType: AddingStruct (0x1025)
    FunctionType: int AddingStruct::(int, int) (0x1040)
    Name: AddingStruct_add
  }
```

**Comment by andrewrk** [comment]
Landed in 6df8e4bca73309f2e340dbfa9031f1bb16a73bcc, thanks @vegecode!

**Comment by emekoi** [comment]
@andrewrk close this?
